third_party/move/move-vm/runtime/src/storage/type_converter.rs

```diff
@@ -0,0 +1,562 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    config::VMConfig,
+    loader::{
+        LegacyModuleStorageAdapter, Loader, PseudoGasContext, MAX_TYPE_TO_LAYOUT_NODES,
+        VALUE_DEPTH_MAX,
+    },
+    storage::{struct_name_index_map::StructNameIndexMap, ty_cache::StructInfoCache},
+    ModuleStorage,
+};
+use move_binary_format::errors::{PartialVMError, PartialVMResult};
+use move_core_types::{
+    language_storage::{StructTag, TypeTag},
+    value::{IdentifierMappingKind, MoveFieldLayout, MoveStructLayout, MoveTypeLayout},
+    vm_status::StatusCode,
+};
+use move_vm_types::loaded_data::runtime_types::{StructLayout, StructNameIndex, StructType, Type};
+use std::sync::Arc;
+
+/// A trait allowing to convert runtime types into other types used throughout the stack.
+#[allow(private_bounds)]
+pub trait TypeConverter: TypeConverterBase {
+    /// Converts a runtime type to a type layout.
+    fn type_to_type_layout(&mut self, ty: &Type) -> PartialVMResult<MoveTypeLayout> {
+        let mut count = 0;
+        self.type_to_type_layout_impl(ty, &mut count, 1)
+            .map(|(l, _)| l)
+    }
+
+    /// Converts a runtime type to a fully annotated type layout, containing information about
+    /// field names.
+    fn type_to_fully_annotated_layout(&self, ty: &Type) -> PartialVMResult<MoveTypeLayout> {
+        let mut count = 0;
+        self.type_to_fully_annotated_layout_impl(ty, &mut count, 1)
+    }
+
+    /// Converts a runtime type to a type tag.
+    fn type_to_type_tag(&self, ty: &Type) -> PartialVMResult<TypeTag> {
```

Comment:
FYI, I have a PR to move type tag conversion outside of loader as well...: https://github.com/aptos-labs/aptos-core/pull/15676. We should probably sync changes for this one.

Reply:
Now integrated, but renamed from 'builder' to 'converter'.